### ISSUE TITLE
Illegal item length failure when reading DICOM

### ISSUE THREAD

**Issue by efahl** [dgobbi/vtk-dicom#186]
Windows-10-10.0.18362-SP0
Python 3.7.4 (tags/v3.7.4:e09359112e, Jul 8 2019, 20:34:20) [MSC v.1916 64 bit (AMD64)]
VTK 8.2.0
Remote/vtkDICOM> git branch --list -v
* master a1105db Rename variables for greater clarity

Possibly related to https://github.com/dgobbi/vtk-dicom/issues/158 ?

The error is
```
ERROR: In Remote/vtkDICOM/Source/vtkDICOMReader.cxx, line 2404
vtkDICOMReader (000002267553E8D0): ERROR: In Remote/vtkDICOM/Source/vtkDICOMParser.cxx, line 2397
vtkDICOMParser (00000226755FB4F0): At byte offset 102116 in file fail2404.anon.dcm: Illegal item length FFFFFFFF encountered.

ERROR: In Remote/vtkDICOM/Source/vtkDICOMReader.cxx, line 2404
vtkDICOMReader (000002267553E8D0): ERROR: In Remote/vtkDICOM/Source/vtkDICOMParser.cxx, line 2397
vtkDICOMParser (00000226755FB4F0): At byte offset 102946 in file fail2404.anon.dcm: Item length exceeds the bytes remaining in file.
```
Code to reproduce, using the attached anonymized DICOM in this zip: [fail.zip](https://github.com/dgobbi/vtk-dicom/files/4072991/fail.zip)

```
import vtk

vo = vtk.vtkFileOutputWindow()
vo.SetFileName('vtk.log')
vo.FlushOn()
vo.SetInstance(vo)


reader = vtk.vtkDICOMReader()
reader.SetFileName('fail2404.anon.dcm')
reader.Update()
```

GDCM seems to read the file just fine (in fact I used gdcmanon to anonymize the file); I can use gdcmconv to change the image encoding and then vtkDICOM can read the file, but in its original form, we get errors these on various images.

Is there any way to enable more diagnostics, so that I could see what element is the actual problem and repair them on the fly as a workaround?  I have not had much luck trying to deduce the root cause of this issue from Python, as my ability to get into the C++ code is quite rudimentary.

**Comment by dgobbi** [comment]
Hi Eric, thanks for attaching the image.  It contain an IconImageSequence (0088,0200) with a compressed icon -- specifically, element (7FE0,0010) within that sequence.  Compressed icons are forbidden by the DICOM standard [(see CP165)](http://www.dclunie.com/dicom-status/status.html#CP165), and they cause problems with some DICOM software (including, apparently, the vtkDICOMReader).  Running these images through gdcmconv to fix them is a good strategy.

There isn't any setting to make vtkDICOMReader more verbose, but I'll expand the error message so that it gives the tag of the data element that caused the error (for cases like this, I'll also need to say what sequence the element was embedded within).

I will also make the reader tolerant of these compressed-icon images, just like gdcm and dcmtk.

**Comment by efahl** [comment]
Thanks a bunch, David, that clarifies things immensely.  I was thinking it was an issue with the top-level PixelData and couldn't figure out why the problem persisted when I simply modified that; never thought to look into the sub-elements for the culprit.

**Comment by malaterre** [comment]
> Compressed icons are forbidden by the DICOM standard [(see CP165)](http://www.dclunie.com/dicom-status/status.html#CP165)

This is incorrect. CP165 simply clarify the encoding (either native or encapsulated) for the Pixel Data element included in the SQ item.

Use a round trip dcmdjpeg/dcmcjpeg (DCMTK) if you do not trust GDCM. gdcmconv by default will not compress the Pixel Data in the icon sequence. But you can use the (legal) switch: `--compress-icon`

**Comment by dgobbi** [comment]
Thanks for the correction, Mathieu.

In this particular file, the icon's encapsulated PixelData had an explicit VR of OW (instead of OB).  So the file was still incorrect. The work-around (970d02c5) was to allow undefined length with OW.

**Comment by dgobbi** [comment]
Fix is in the latest release (v0.8.13).
